clang/test/SemaCXX/friend.cpp

```diff
@@ -429,3 +429,19 @@ namespace qualified_friend_no_match {
     friend void Y::f(double); // expected-error {{friend declaration of 'f' does not match any declaration in 'qualified_friend_no_match::Y'}}
   };
 }
+
+namespace gh21483 {
+template <typename I>
+struct B {
+  struct mixin {
+    int type;
+    friend void foo(B<I>::mixin it) {
+      (void)it.mixin::type;
```

Comment:
Why is `it.mixin` not an error? I thought `it` was of type `B<int>::mixin`, which has no `mixin` field.

Reply:
`::` is scope resolution operator.
A more clear example is 
```c++
struct T {
    int a;
};
struct T1 : public T {
    int a;
};

void foo () {
    T1{}.T::a;
    T1{}.T1::a;
}
```
https://godbolt.org/z/K1b1jodse

Reply:
Ah I see it now after playing around with it a bit. In this case line 439 could have accessed the same value through `(void)it.type` since `mixin` doesn't inherit anything, but `it.mixin::type` is allowed regardless.